internal/service/gateway/tunnel.go

```diff
@@ -94,14 +99,106 @@ func ResourceTunnel() *schema.Resource {
 				Elem:        ipsecPropertiesSchema(),
 			},
 		},
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Type:    resourceTunnelResourceV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourceTunnelStateUpgradeV0,
+				Version: 0,
+			},
+		},
 	}
 }
 
+func resourceTunnelResourceV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description:      "The name of the tunnel, should be unique within the connection",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validateName,
+			},
+			"connection_id": {
+				Description: "ID of the upcloud_gateway_connection resource to which the tunnel belongs",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"local_address_name": {
+				Description:      "Public (UpCloud) endpoint address of this tunnel",
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateDiagFunc: validateName,
+			},
+			"remote_address": {
+				Description: "Remote public IP address of the tunnel",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"operational_state": {
+				Description: "Tunnel's current operational, effective state",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"ipsec_auth_psk": {
+				Description: "Configuration for authenticating with pre-shared key",
+				ForceNew:    true,
+				Type:        schema.TypeList,
+				Required:    true,
+				MaxItems:    1,
+				Elem:        ipsecAuthPSKSchema(),
+			},
+			"ipsec_properties": {
+				Description: "IPsec configuration for the tunnel",
+				Type:        schema.TypeList,
+				Optional:    true,
+				Computed:    true,
+				MaxItems:    1,
+				Elem:        ipsecPropertiesSchema(),
+			},
+		},
+	}
+}
+
+func resourceTunnelStateUpgradeV0(ctx context.Context, rawState map[string]any, meta any) (map[string]any, error) {
+	var (
+		svc            = meta.(*service.Service)
+		serviceUUID    string
+		connectionUUID string
+		name           string
+	)
+
+	if err := utils.UnmarshalID(rawState["id"].(string), &serviceUUID, &connectionUUID, &name); err != nil {
+		return rawState, err
+	}
+
+	tunnels, err := svc.GetGatewayConnectionTunnels(ctx, &request.GetGatewayConnectionTunnelsRequest{
+		ServiceUUID:    serviceUUID,
+		ConnectionUUID: connectionUUID,
+	})
+	if err != nil {
+		return rawState, err
+	}
+
+	for _, tunnel := range tunnels {
+		if tunnel.Name == rawState["name"].(string) {
+			rawState["uuid"] = tunnel.UUID
+
+			return rawState, nil
+		}
+	}
+
+	return rawState, fmt.Errorf("tunnel by name %s not found", name)
+}
+
 func resourceTunnelCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) (diags diag.Diagnostics) {
```

Comment:
Here also name is used in the id instead of UUID 🤔 Is changing the id possible in the upgrader?

Reply:
Changed to use the UUID, better for import purposes.